pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -0,0 +1,429 @@
+	for _, role := range roles {
+		requiredBindings[role] = cloudresourcemanager.Binding{
+			Members: []string{"serviceAccount:" + members},
+			Role:    role,
+		}
+	}
+	return requiredBindings
+}
+
+// NewGCPSecretCR returns a Secret CR formatted for GCP
+func NewGcpSecretCR(namespace, creds string) *corev1.Secret {
+	return &corev1.Secret{
+		Type: "Opaque",
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      gcpSecretName,
+			Namespace: namespace,
+		},
+		Data: map[string][]byte{
+			"osServiceAccount.json": []byte(creds),
+		},
+	}
+}
+
+// AddOrUpdateBinding checks if a binding from a map of bindings whose keys are the binding.Role exists in a list and if so it appends any new members to that binding.
+// If the required binding does not exist it creates a new binding for the role
+// it returns a []*cloudresourcemanager.Binding that contains all the previous bindings and the new ones if no new bindings are required it returns false
+func AddOrUpdateBinding(existingBindings []*cloudresourcemanager.Binding, requiredBindings map[string]cloudresourcemanager.Binding) (bool, []*cloudresourcemanager.Binding) {
+	Modified := false
+
+	for _, eBinding := range existingBindings {
+		if rBinding, ok := requiredBindings[eBinding.Role]; ok {
+			// check if members list contains from existing contains members from required
+			for _, rMember := range rBinding.Members {
+				if !stringInSlice(rMember, eBinding.Members) {
+					Modified = true
+					// If required member is not in existing member list add it
+					eBinding.Members = append(eBinding.Members, rMember)
+				}
+			}
+			// delete processed key from requiredBindings
+			delete(requiredBindings, eBinding.Role)
+		}
+
+	}
+
+	// take the remaining bindings from map of required bindings and append it to the list of existing bindings
+	if len(requiredBindings) > 0 {
+		Modified = true
+		for _, binding := range requiredBindings {
+			existingBindings = append(existingBindings, &binding)
+		}
+	}
+
+	return Modified, existingBindings
+}
+
+func stringInSlice(a string, list []string) bool {
+	for _, b := range list {
+		if b == a {
+			return true
+		}
+	}
+	return false
+}
+
+// TODO(Raf) Clean serviceAccount from member in bindings
+
+func findMemberIndex(searchMember string, members []string) int {
+	for index, value := range members {
+		if value == searchMember {
+			return index
+		}
+	}
+	return -1
+}
+
+// remove removes a given index from a []string
+func remove(slice []string, s int) []string {
+	return append(slice[:s], slice[s+1:]...)
+}
+
+// SecretExists returns a boolean to the caller basd on the secretName and namespace args.
+func SecretExists(kubeClient client.Client, secretName, namespace string) bool {
+
+	s := &corev1.Secret{}
+
+	err := kubeClient.Get(context.TODO(), kubetypes.NamespacedName{Name: secretName, Namespace: namespace}, s)
```

Comment:
check error type here with `apimachinery errors` package to check it this is notFound error

Reply:
What should we be doing in the case the error is not notFound ? Should since the function only returns a bool I would expect it to turn false no matter what the error is .